src/foam/util/AddressUtil.java

```diff
@@ -57,4 +66,73 @@
     return new String[] { streetNumber, streetName };
   }
 
+  public static String normalizeRegion(X x, String country, String regionCode) {
+    String normalizedRegion = country + "-" + regionCode;
+    DAO regionDAO = (DAO) x.get("regionDAO");
+    Region found = (Region) regionDAO.find(regionCode);
+    if ( found != null )
+      return found.getCode();
+
+    ArraySink sink = (ArraySink) regionDAO.where(AND(
+      EQ(Region.COUNTRY_ID, country),
+      OR(
+        EQ(Region.ISO_CODE, regionCode),
+        STARTS_WITH_IC(Region.NAME, regionCode)
+      )
+    )).select(new ArraySink());
+
+    List<Region> regions = sink.getArray();
+    
+    // Take the first match as the default
+    if (regions.size() > 0)
+      normalizedRegion = regions.get(0).getCode();
+    
+    // When there are more than one, take the first matches on the ISO code or on the full name (i.e. not just CONTAINS_IC)
+    if (regions.size() > 1) {
+      for ( int i = 0; i < regions.size(); i++ ) {
+        Region region = (Region) regions.get(i);
+        if ( region.getIsoCode().equals(regionCode) || region.getName().equalsIgnoreCase(regionCode) ) {
+          normalizedRegion = region.getCode();
+          break;
+        }
+      }
+    }
+    
+    return normalizedRegion;
+  }
+
+  public static String normalizeCountry(X x, String countryCode) {
+    String normalizedCountry = countryCode;
+    DAO countryDAO = (DAO) x.get("countryDAO");
+    Country found = (Country) countryDAO.find(countryCode);
+    if ( found != null )
+      return found.getCode();
+
+    ArraySink sink = (ArraySink) countryDAO
```

Comment:
Rather than selecting into an ArraySink, you could just process directly. Then you can stop processing once you find the correct value and you don't generate garbage.
See the following example taken from Group.js:

 .select(new AbstractSink() {
              public void put(Object o, Detachable d) {
                GroupPermissionJunction j = (GroupPermissionJunction) o;

                if ( j.getTargetId().isBlank() ) return;

                if ( j.getTargetId().startsWith("@") ) {
                  Group group = (Group) dao.find(j.getTargetId().substring(1));

                  if ( group != null && group.implies(x, permission) ) {
                    granted[0] = true;
                    d.detach();
                  }
                } else if ( new AuthPermission(j.getTargetId()).implies(permission) ) {
                  granted[0] = true;
                  d.detach();
                }
              }
            });

The d.detach() stops the select() early.